en/advanced/ai.md

```diff
@@ -0,0 +1,79 @@
+# AI functionality in JabRef
+
+## Chat model
+
+**Type**: enumeration
+
+**Requirements**: choose one available from combo box
+
+Chat model specifies what AI models can you use. This will differ from one provider to other. Models vary in their accuracy, knowledge of the world, context window (what amount of information can they process).
+
+Currently only OpenAI models are supported.
+
+## Embedding model
+
+**Type**: enumeration
+
+**Requirements**: choose one available from combo box
+
+Embedding model transforms a document (or a piece of text) into a vector (an ordered collection of numbers). This is used to supply the AI with relevant information regarding your questions.
+
+Different embedding models have different performance: this includes accuracy and how fast embeddings can be computed. `Q` at the end of model name usually means *quantized* (meaning *reduced*, *simplified*). These models are fast, but provide less accuracy.
+
+Currently only local embedding models are supported. That means you don't have to provide a new API key and all the logic will be run on your machine. 
+
+## Instruction
+
+**Type**: string
+
+**Requirements**: not empty
+
+An instruction (also known as "system message") in a Large Language Models (LLMs) sets the tone and rules for the conversation. Think of it as instructions given to the AI before it starts interacting with a user. It guides the AI on how to respond, ensuring it stays on topic and behaves appropriately. For example, a system message might tell the AI to be formal, concise, or provide detailed explanations. This helps the AI provide more relevant and useful answers tailored to the specific needs of the user.
+
+## Message window size
+
+**Type**: integer
+
+**Requirements**: > 0
+
+The "message window size" in a Large Language Model (LLM) refers to the number of recent messages or interactions that the model remembers during a conversation. This parameter determines how much context the LLM considers when generating responses.
```

Comment:
See https://docs.gpt4all.io/gpt4all_desktop/settings.html#sampling-settings for different explanation of settings:
![image](https://github.com/JabRef/user-documentation/assets/73715071/1e984701-8b0f-4f99-b5c7-6db0c48a4225)
That documentation is not completely correct either. Context window is measured in tokens. More specifically, it is the sum of tokens of system prompt + user prompt + chunks/snippets of embeddings that were added to the user prompt + the responses by the model.

Reply:
Explanation of model settings by artefact2: https://artefact2.github.io/llm-sampling/index.xhtml

Reply:
[artefact2.github.io/llm-sampling/index.xhtml](https://artefact2.github.io/llm-sampling/index.xhtml) might be a good explanation, but I think it's too technical and mathematically inclined. And I haven't included those parameters to the JabRef (except temperature)

Reply:
"Message window size" is the same as "context length" in GPT4All, as far as I understand it and this paragraph should be rewritten. The description of Message window size is not accurate.

Reply:
Or is it really the number of messages?

Reply:
in `langchain4j` it's really message number. It handles the context length of the chat history somehow on it's own

Reply:
But if a user message has 50 tokens or if a user message has 5000 tokens. Will both be counted as one? Models have a maximum context window size they are trained on and those are counted by tokens. A model responses degrade steeply, if going above that limit, regardless if it is in the middle of the sentence or not. It depends on the amount of tokens. It does not make sense to count by number of messages, as each message can have different amounts of tokens.

Reply:
The algorithms for managing chat messages in `langchain4j` are very messy, so I decided not to touch them at all.

I've double checked the algorithm that `langchain4j` uses in `MessageWindowChatMemory`:
1. It tries to estimate the size of chat history
2. When it's overflowed it removes several old messages to conform to context size requirements

Reply:
Ok, Message Window Size is derived from LangChain specific code related to MessageWindowChatMemory`. I tried finding "Message Window Size" in their documentation, but could not. What I found was https://docs.langchain4j.dev/tutorials/chat-memory/. We basically have now made up our own name for this behaviour. It is fine, but we should take care to not confuse it with "Context Window Size" of an LLM. Both are not the same. As I understand it, our "Message Window Size" will remove messages only, if it overflows, regardless, if users have already exceeded the models context window size long ago. Please correct me, if I am wrong.

This deserves a tiny rewrite and I will resolve this, as soon as we have disentangled this.